src/Datasource/Exception/InvalidPrimaryKeyException.php

```diff
@@ -17,12 +17,16 @@
 namespace Cake\Datasource\Exception;
 
 use Cake\Core\Exception\Exception;
+use Cake\Core\Exception\HttpExceptionCodeInterface;
+use Cake\Core\Exception\HttpExceptionCodeTrait;
 
 /**
  * Exception raised when the provided primary key does not match the table primary key
  */
-class InvalidPrimaryKeyException extends Exception
+class InvalidPrimaryKeyException extends Exception implements HttpExceptionCodeInterface
```

Comment:
The dbal throwing exceptions with HTTP status code feels pretty weird.

IMO a proper way to handle this would be to remove the 4xx/5xx exception codes from all exceptions exception the ones under http package. The exception renderer should then wrap all non http exception with a http exception with appropriate status code.

Reply:
That does seem cleaner. I don't think any non-http exceptions would need a special http code (just 404 or 500).

I'll wait for other feedback before implementing that since this PR covers a few of the recommendations in the issue.